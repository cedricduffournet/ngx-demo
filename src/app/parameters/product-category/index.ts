export * from './parameter-product-category.component';
