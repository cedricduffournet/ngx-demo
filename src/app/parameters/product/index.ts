export * from './parameter-product.component';
export * from './parameter-product-add.component';
export * from './parameter-product-details.component';
