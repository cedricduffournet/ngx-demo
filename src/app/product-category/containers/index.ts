export * from './product-category-add-modal/product-category-add-modal.component';
export * from './product-category-update-modal/product-category-update-modal.component';
export * from './product-category-delete-modal/product-category-delete-modal.component';
export * from './product-category-list-view/product-category-list-view.component';
