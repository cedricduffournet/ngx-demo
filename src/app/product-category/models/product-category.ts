import { schema } from 'normalizr';

export interface ProductCategory {
  id: number;
  name: string;
}

export const productCategorySchema = new schema.Entity('productCategories');
