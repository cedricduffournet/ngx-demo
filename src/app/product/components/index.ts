export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
