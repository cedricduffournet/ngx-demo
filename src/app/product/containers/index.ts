export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
