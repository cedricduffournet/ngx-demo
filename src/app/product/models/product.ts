import { schema } from 'normalizr';
import {
  ProductCategory,
  productCategorySchema
} from '@app/product-category/models/product-category';

export interface ProductBase {
  id: number;
  name: string;
  description: string;
  priceAmount: number;
}

export interface Product extends ProductBase {
  categories: number[];
}

export interface ProductFull extends ProductBase {
  categories: ProductCategory[];
}

export const productSchema = new schema.Entity('products', {
  categories: [productCategorySchema]
});

export const getPrice = (price: number) => {
  if (price > 0) {
    return price / 100;
  }
  return 0;
};
