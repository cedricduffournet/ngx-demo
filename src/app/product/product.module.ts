import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent
} from '@app/product/components';
import {
  ProductAddModalComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddModalComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductAddModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent],
  providers: [ProductFacade]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
