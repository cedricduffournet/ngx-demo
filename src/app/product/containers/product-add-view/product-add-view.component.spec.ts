import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { Router } from '@angular/router';

import { provideMockStore } from '@ngrx/store/testing';
import { of } from 'rxjs';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { TranslateModule } from '@ngx-translate/core';

import { FormModule } from '@app/shared/form';
import { PageInnerModule } from '@app/shared/page-inner';
import { ValidationActionModule } from '@app/shared/validation-action';

import { ProductAddViewComponent } from '@app/product/containers';
import {
  ProductAddComponent,
  ProductFormComponent
} from '@app/product/components';
import { Product } from '@app/product/models/product';
import { ProductFacade } from '@app/product/state/product.facade';
import { ProductCategoryFacade } from '@app/product-category/state/product-category.facade';

describe('ProductAddModalComponent', () => {
  let fixture: ComponentFixture<ProductAddViewComponent>;
  let component: ProductAddViewComponent;
  let facade: ProductFacade;
  let router: Router;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        ProductAddComponent,
        ProductAddViewComponent,
        ProductFormComponent
      ],
      imports: [
        TranslateModule.forRoot(),
        FormModule,
        ValidationActionModule,
        ReactiveFormsModule,
        PageInnerModule
      ],
      providers: [
        provideMockStore(),
        BsModalRef,
        {
          provide: ProductCategoryFacade,
          useValue: {
            productCategories$: of([]),
            loadProductCategories: jest.fn()
          }
        },
        {
          provide: ProductFacade,
          useValue: {
            added$: of(false),
            addProduct: jest.fn()
          }
        },
        {
          provide: Router,
          useValue: { navigate: jest.fn() }
        }
      ]
    });
    fixture = TestBed.createComponent(ProductAddViewComponent);
    facade = TestBed.get(ProductFacade);
    router = TestBed.get(Router);
    component = fixture.componentInstance;
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should call addProduct event on submit', () => {
    spyOn(facade, 'addProduct');
    const product = {
      name: 'name'
    } as Product;
    fixture.detectChanges();
    component.onAdd(product);

    expect(facade.addProduct).toHaveBeenCalledWith(product);
  });

  it('should navigate to list on cancel', () => {
    fixture.detectChanges();
    component.onCancel();
    expect(router.navigate).toHaveBeenCalledWith(['parameters/products']);
  });
});
