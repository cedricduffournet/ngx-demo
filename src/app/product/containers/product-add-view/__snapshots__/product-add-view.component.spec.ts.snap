// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductAddModalComponent should be created 1`] = `
<app-product-add-view
  added$={[Function Observable]}
  adding$="undefined"
  categories$={[Function Observable]}
  facade={[Function Object]}
  loadingCategories$="undefined"
  productCategoryFacade={[Function Object]}
  router={[Function Object]}
>
  <app-page-inner>
    <div
      class="px-3 py-3 "
    >
      <h2
        ng-reflect-translate=""
        translate=""
      >
        ADD_PRODUCT_TITLE
      </h2>
      
    </div>
  </app-page-inner>
</app-product-add-view>
`;
